edx_repo_tools/codemods/node16/gha_ci_modernizer.py

```diff
@@ -0,0 +1,74 @@
+"""
+Django Matrix Modernizer for Github Actions CI
+"""
+import click
+from copy import deepcopy
+
+from edx_repo_tools.utils import YamlLoader
+
+ALLOWED_NODE_VERSIONS = [12, 14, 16]
+ALLOWED_NPM_VERSION = '8.x.x'
+
+
+class GithubCINodeModernizer(YamlLoader):
+    def __init__(self, file_path):
+        super().__init__(file_path)
+        self.yml_instance.default_flow_style = None
+        self.yml_instance.indent(mapping=2, sequence=2, offset=0)
+
+    def _add_or_update_strategy_matrix(self, job_name):
+        if 'strategy' in self.elements['jobs'][job_name] and 'matrix' in self.elements['jobs'][job_name]['strategy']:
+            self.elements['jobs'][job_name]['strategy']['matrix']['node'] = ALLOWED_NODE_VERSIONS
+        else:
+            self.elements['jobs'][job_name]['strategy'] = {'matrix': {'node': ALLOWED_NODE_VERSIONS}}
+            self.elements['jobs'][job_name].move_to_end('strategy', last=False)
+            self.elements['jobs'][job_name].move_to_end('runs-on', last=False)
+            if 'name' in self.elements['jobs'][job_name]:
+                self.elements['jobs'][job_name].move_to_end('name', last=False)
+
+    def _update_node_version(self, job_name, step):
+        step_index = self.elements['jobs'][job_name]['steps'].index(step)
+        self.elements['jobs'][job_name]['steps'][step_index]['with']['node-version'] = '${{ matrix.node }}'
+
+    def _update_npm_version(self, job_name, step):
+        step_index = self.elements['jobs'][job_name]['steps'].index(step)
+        self.elements['jobs'][job_name]['steps'][step_index]['run'] = 'npm i -g npm@'+ALLOWED_NPM_VERSION
+
+    def _update_job_steps(self, job_name, job):
+        steps = job.get('steps')
+        if not steps:
+            return
+        for step in steps:
+            if 'name' in step and step['name'] == 'Setup Nodejs':
+                self._update_node_version(job_name, step)
+            elif 'name' in step and step['name'] == 'Setup npm':
+                self._update_npm_version(job_name, step)
+
+    def _update_job_name(self, job_name, job):
+        self.elements['jobs']['tests'] = deepcopy(job)
+        self.elements['jobs'].move_to_end('tests', last=False)
+        self.elements['jobs'].pop(job_name)
+
+    def _update_job(self):
+        for job_name, job in self.elements.get('jobs').items():
+            self._add_or_update_strategy_matrix(job_name)
+            self._update_job_steps(job_name, job)
+            if job_name != 'tests':
```

Comment:
If there are multiple jobs in a workflow then we are updating all the job names with the same name? Wouldn't it be better to restrict ourselves to the only name we want to update.